lib/scala/project-manager/src/main/scala/org/enso/projectmanager/requesthandler/RequestHandler.scala

```diff
@@ -0,0 +1,106 @@
+package org.enso.projectmanager.requesthandler
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Cancellable, Stash, Status}
+import akka.pattern.pipe
+import org.enso.jsonrpc.Errors.ServiceError
+import org.enso.jsonrpc.{Id, Method, Request, ResponseError}
+import org.enso.projectmanager.control.effect.Exec
+import org.enso.projectmanager.service.versionmanagement.ProgressNotification
+import org.enso.projectmanager.service.versionmanagement.ProgressNotification.translateProgressNotification
+import org.enso.projectmanager.util.UnhandledLogging
+
+import scala.concurrent.duration.FiniteDuration
+
+/** A helper class that gathers common request handling logic.
+  *
+  * It manages timeouts and sending the request result (in case of success but
+  * also failure or timeout).
+  *
+  * @param handledMethod method that this handler deals with; used in logging
+  *                      and to relate progress updates to the method
+  * @param requestTimeout timeout for the request; if set, the request will be
+  *                       marked as failed after the specified time; the request
+  *                       logic is however NOT cancelled as this is not possible
+  *                       to do in a general way
+  */
+abstract class RequestHandler[
+  F[+_, +_]: Exec,
+  FailureType: FailureMapper: Manifest
+](
+  handledMethod: Method,
+  requestTimeout: Option[FiniteDuration]
+) extends Actor
+    with ActorLogging
+    with Stash
+    with UnhandledLogging {
+  override def receive: Receive = requestStage
+
+  import context.dispatcher
+
+  /** Waits for the request, tries to pass it into the [[handleRequest]]
+    * function, sets up the timeout and routing of the result.
+    */
+  private def requestStage: Receive = {
+    val composition: Any => Option[Unit] = {
+      case request @ Request(_, id, _) =>
+        val result = handleRequest.lift(request)
+        result.map { f =>
+          Exec[F].exec(f).pipeTo(self)
+          val cancellable = {
+            requestTimeout.map { timeout =>
+              context.system.scheduler.scheduleOnce(
+                timeout,
+                self,
+                RequestTimeout
+              )
+            }
+          }
+          context.become(responseStage(id, sender(), cancellable))
+        }
+      case _ => None
+    }
+    Function.unlift(composition)
```

Comment:
Because this is not an ordinary partial function.

I take advantage of that that I can take a look at the input, do stuff and *then* decide if I want to return a `Some` (so decide to be defined) or `None`.

I use that in the first branch, where I map over `result` which is applying the inner partial function.

Now, the inner partial function (`handleRequest`) may or may not be defined so we may need to handle both of these cases.

-----------

But after having a look I decided to not be afraid of types and make the request handler type-safe (as the function was partial not because it had to be, but because I could not guarantee proper types).

Now the implementations are even simpler at the cost of making the types in the inheritance a bit more complicated - but I think it is worth it.